Neos.Flow/Classes/Persistence/Doctrine/PersistenceManager.php

```diff
@@ -86,23 +93,44 @@ public function injectLogger(LoggerInterface $logger)
      */
     public function persistAll($onlyAllowedObjects = false)
     {
-        if ($onlyAllowedObjects) {
-            $unitOfWork = $this->entityManager->getUnitOfWork();
-            /** @var \Doctrine\ORM\UnitOfWork $unitOfWork */
-            $unitOfWork->computeChangeSets();
+        if (!$this->entityManager->isOpen()) {
+            $this->logger->error('persistAll() skipped flushing data, the Doctrine EntityManager is closed. Check the logs for error message.', LogEnvironment::fromMethodName(__METHOD__));
+            return;
+        }
+
+        $this->onlyAllowedObjects = $onlyAllowedObjects;
+        $this->entityManager->flush();
+        $this->onlyAllowedObjects = false;
+        $this->emitAllObjectsPersisted();
+    }
+
+    /**
+     * Doctrine onFlush listener that checks for only allowed objects and reconnects
+     * if the database connection was closed.
+     *
+     * @param OnFlushEventArgs $args
+     * @throws PersistenceException
+     */
+    public function onFlush(OnFlushEventArgs $args)
+    {
+        $unitOfWork = $args->getEntityManager()->getUnitOfWork();
+        if ($unitOfWork->getScheduledEntityInsertions() === []
+            && $unitOfWork->getScheduledEntityUpdates() === []
+            && $unitOfWork->getScheduledEntityDeletions() === []
+            && $unitOfWork->getScheduledCollectionDeletions() === []
+            && $unitOfWork->getScheduledCollectionUpdates() === []
+        ) {
+            return;
+        }
+
+        if ($this->onlyAllowedObjects) {
             $objectsToBePersisted = $unitOfWork->getScheduledEntityUpdates() + $unitOfWork->getScheduledEntityDeletions() + $unitOfWork->getScheduledEntityInsertions();
             foreach ($objectsToBePersisted as $object) {
                 $this->throwExceptionIfObjectIsNotAllowed($object);
             }
         }
 
-        if (!$this->entityManager->isOpen()) {
-            $this->logger->error('persistAll() skipped flushing data, the Doctrine EntityManager is closed. Check the logs for error message.', LogEnvironment::fromMethodName(__METHOD__));
-            return;
-        }
-
-        /** @var Connection $connection */
-        $connection = $this->entityManager->getConnection();
```

Comment:
This part was not guarded by a check that there are actual changes in the UnitOfWork. Simply calling `$this->hasUnpersistedChanges()` would have invoked the costly `computeChangesets()` twice, or even thrice for safe requests (onlyAllowedObjects = true). With this change, it will always only be invoked once, no matter what and the database reconnect only if there are changes.